src/manifolds/Sphere.jl

```diff
@@ -274,93 +342,98 @@ end
 end
 
 @doc raw"""
-    project(M::Sphere, p)
+    project(M::AbstractSphere, p)
 
 Project the point `p` from the embedding onto the [`Sphere`](@ref) `M`.
 
 ````math
-\operatorname{proj}_{𝕊^n}(p) = \frac{p}{\lVert p \rVert_2}.
+    \operatorname{proj}(p) = \frac{p}{\lVert p \rVert},
 ````
+where $\lVert\cdot\rVert$ denotes the usual 2-norm for vectors if $m=1$ and the Frobenius
+norm for the case $m>1$.
 """
-project(::Sphere, ::Any)
+project(::AbstractSphere, ::Any)
 
-project!(S::Sphere, q, p) = copyto!(q, p./ norm(p))
+project!(S::AbstractSphere, q, p) = copyto!(q, p./ norm(p))
 
 @doc raw"""
-    project(M::Sphere, p, X)
+    project(M::AbstractSphere, p, X)
 
 Project the point `X` onto the tangent space at `p` on the [`Sphere`](@ref) `M`.
 
 ````math
 \operatorname{proj}_{p}(X) = X - ⟨p, X⟩p
 ````
 """
-project(::Sphere, ::Any, ::Any)
+project(::AbstractSphere, ::Any, ::Any)
 
-project!(S::Sphere, Y, p, X) = (Y .= X .- dot(p, X) .* p)
+project!(S::AbstractSphere, Y, p, X) = (Y .= X .- real(dot(p, X)) .* p)
 
 @doc raw"""
-    representation_size(M::Sphere)
+    representation_size(M::AbstractSphere)
 
-Return the size points on the [`Sphere`](@ref) `M` are represented as, i.e.
-for the `n`-dimensional [`Sphere`](@ref) it is vectors of size `(n+1,)`.
+Return the size points on the [`AbstractSphere`](@ref) `M` are represented as, i.e., the
+representation size of the embedding.
 """
-@generated representation_size(::Sphere{N}) where {N} = (N + 1,)
+@generated representation_size(M::ArraySphere{N}) where {N} = size_to_tuple(N)
+@generated representation_size(M::Sphere{N}) where {N} = (N+1,)
+
 
 @doc raw"""
-    retract(M::Sphere, p, X, ::ProjectionRetraction)
+    retract(M::AbstractSphere, p, X, ::ProjectionRetraction)
 
 Compute the retraction that is based on projection, i.e.
 
 ````math
 \operatorname{retr}_p(X) = \frac{p+X}{\lVert p+X \rVert_2}
 ````
 """
-retract(::Sphere, ::Any, ::Any, ::ProjectionRetraction)
+retract(::AbstractSphere, ::Any, ::Any, ::ProjectionRetraction)
 
-function retract!(M::Sphere, q, p, X, ::ProjectionRetraction)
+function retract!(M::AbstractSphere, q, p, X, ::ProjectionRetraction)
     q .= p .+ X
     return project!(M, q, q)
 end
 
-show(io::IO, ::Sphere{N}) where {N} = print(io, "Sphere($(N))")
+show(io::IO, ::Sphere{n,𝔽}) where {n,𝔽} = print(io, "Sphere($(n); field = $(𝔽))")
+show(io::IO, ::ArraySphere{N,𝔽}) where {N,𝔽} = print(io, "ArraySphere($(join(N.parameters, ", ")); field = $(𝔽))")
 
 """
-    uniform_distribution(M::Sphere, p)
+    uniform_distribution(M::AbstractSphere, p)
 
-Uniform distribution on given [`Sphere`](@ref) `M`. Generated points will be of
+Uniform distribution on given [`AbstractSphere`](@ref) `M`. Generated points will be of
 similar type as `p`.
 """
-function uniform_distribution(M::Sphere, p)
+function uniform_distribution(M::AbstractSphere, p)
     d = Distributions.MvNormal(zero(p), 1.0)
```

Comment:
Hmm, this won't generalize beyond the 1D case. You could also extend to the 2D case using `MatrixNormal` as in #142, but to go beyond that with `ProjectedPointDistribution` you'll need something else. I don't think Distributions supports multidimensional array distributions.

Reply:
You could always take the stand normal implementation in #142 and generalize it for multidimensional arrays.

Reply:
Then we should for this PR maybe just restrict it to the `Sphere` again? Does it work for the complex case?

Once #142 is done we can generalise it to the `ArraySphere`, too.

Reply:
That's the other thing. I don't think Distributions.jl supports any complex distributions currently. These gaps are getting annoying enough that it's probably time to start implementing some of #57, at least the `Uniform` distribution for starters. I'd say limit to real `Sphere` for now, and once I've made a PR for `Uniform`, we can switch it over.

Reply:
Thanks for noticing, I will restrict it.